src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Utilities/NullableHelpers/NullableHelpers.cs

```diff
@@ -0,0 +1,109 @@
+            }
+
+            return IsSymbolAssignedPossiblyNullValue(semanticModel, rootOperation, declaredType) == true;
+        }
+
+        /// <summary>
+        /// Given an operation, goes through all decendent operations and returns true if the symbol passed in
+        /// is ever assigned a possibly null value as determined by nullable flow state. Returns
+        /// null if no references are found, letting the caller determine what to do with that information
+        /// </summary>
+        public static bool? IsSymbolAssignedPossiblyNullValue(SemanticModel semanticModel, IOperation operation, ISymbol symbol)
+        {
+            var references = operation.DescendantsAndSelf()
+                .Where(o => IsSymbolReferencedByOperation(o, symbol));
+
+            var hasReference = false;
+
+            foreach (var reference in references)
+            {
+                hasReference = true;
+
+                // foreach statements are handled special because the iterator is not assignable, so the elementtype 
+                // annotation is accurate for determining if the loop declaration has a reference that allows the symbol
+                // to be null
+                if (reference is IForEachLoopOperation forEachLoop)
+                {
+                    var foreachInfo = semanticModel.GetForEachStatementInfo((CommonForEachStatementSyntax)forEachLoop.Syntax);
+
+                    if (foreachInfo.ElementType is null)
+                    {
+                        continue;
+                    }
+
+                    // Use NotAnnotated here to keep both Annotated and None (oblivious) treated the same, since
+                    // this is directly looking at the annotation and not the flow state
+                    if (foreachInfo.ElementType.NullableAnnotation != NullableAnnotation.NotAnnotated)
+                    {
+                        return true;
+                    }
+
+                    continue;
+                }
+
+                var syntax = reference is IVariableDeclaratorOperation variableDeclarator
+                    ? variableDeclarator.GetVariableInitializer()!.Value.Syntax
+                    : reference.Syntax;
+
+                var typeInfo = semanticModel.GetTypeInfo(syntax);
+
+                if (typeInfo.Nullability.FlowState == NullableFlowState.MaybeNull)
+                {
+                    return true;
+                }
+            }
+
+            return hasReference ? (bool?)false : null;
+        }
+
+        private static bool IsSymbolReferencedByOperation(IOperation operation, ISymbol symbol)
+            => operation switch
+            {
+                ILocalReferenceOperation localReference => localReference.Local.Equals(symbol),
+                IParameterReferenceOperation parameterReference => parameterReference.Parameter.Equals(symbol),
+                IAssignmentOperation assignment => IsSymbolReferencedByOperation(assignment.Target, symbol),
```

Comment:
why do we need thsi?  isn't the caller already descending all iops?  so why do we need to manually recurse ourselves?

Reply:
It depends on what you want to consider as the important operation. Does an `IAssignmentOperation` ever reference a symbol directly or is it the child operations that do? 

If we only take the lowest operation that actually references the symbol we'd just do the inverse: instead of having the information that it's an `IAssignmentOperation` it would be something like an `ILocalReferenceOperation` and then we'd have to check up the tree for the context

Reply:
this would be a valuable comment for the code :)